### PR TITLE
ramips: Add support for MikroTik RouterBOARD RBM11g

### PR DESCRIPTION
**Pull request by TobleMiner**
This commit adds support for the MikroTik RouterBOARD RBM11g.

# Hardware

The RBM11g is a mt7621 based device featuring one GbE port and one
miniPCIe slot with a sim card socket and USB 2.0.

## Switch

The single onboard Ethernet port is connected directly to the CPU. The internal switch is disabled.

## Flash

The device has one spi nor flash chip. It is a 128 Mbit winbond 25Q128FVS
connected to CS0.

## PCIe

The board features a single miniPCIe slot. It has a dedicated mini SIM
socket and a USB 2.0 port. Power to the miniPCIe slot is controlled via
GPIO9.

## USB

There are no external USB ports.

## Power

The board can accept both, passive PoE and external power via a 2.1 mm
barrel jack (center-positive). The input voltage range is 11-32 V.

## Serial port

The device does have an onboard UART on an unpopulated header next to the
flash chip:

```
GND: pin 2
 TX: pin 7
 RX: pin 6
```

Settings: 115200, 8N1

See below illustration for positioning of the header.

```
0 = screw hole
* = some pin
T = TX  pin
R = RX  pin
G = GND pin
```

Pinout:
```
+---------------
|O
|             __
|            /  \
|            \__/
|
|
|
|               +---+
|               |RAM|
| +--+          |   |
| |**|  <- unpopulated header with UART
| |*T|          +---+
| |R*|        +--------+
| |**|        |        |
| |G*|        |  CPU   |
| +--+        |        |
|    +--+     |        |
|    |  |     +--------+
|    +--+  <- flash chip
|O
|       +-----+
|       |     |
|+--+   |     |
||  |   |     |
+---------------------
```

# Installation

To install an OpenWRT image to the device two components must be built:

1. A openwrt initramfs image
2. A openwrt sysupgrade image

## initramfs & sysupgrade image

Select target devices "Mikrotik RBM11G" in
openwrt menuconfig and build the images. This will create the images
"openwrt-ramips-mt7621-mikrotik_rbm11g-initramfs-kernel.bin" and
"openwrt-ramips-mt7621-mikrotik_rbm11g-squashfs-sysupgrade.bin" in the
output directory.

## Installing

**Make sure to back up your RouterOS license in case you do ever want to
go back to RouterOS using "/system license output" and back up the
created license file.**

When rebooted the board will try booting via ethernet first. If your
board does not boot via ethernet automatically you will have to attach
to the serial port and set ethernet as boot device within RouterBOOT.

1. Set up a dhcp server that points the bootfile to tftp server serving
   the "openwrt-ramips-mt7621-mikrotik_rbm11g-initramfs-kernel.bin"
   initramfs image
2. Connect to ethernet port on board
3. Power on the board
4. Wait for OpenWrt to boot

Right now OpenWrt will be running with a SSH server listening. Now
OpenWrt must be flashed to the devices flash:

1. Copy "openwrt-ramips-mt7621-mikrotik_rbm11g-squashfs-sysupgrade.bin"
   to the device using scp.
2. Write openwrt to flash using "sysupgrade
   openwrt-ramips-mt7621-mikrotik_rbm11g-squashfs-sysupgrade.bin"

Once the flashing completes the board will reboot. Disconnect from the
devices ethernet port or stop the DHCP/TFTP server to prevent the device
from booting via ethernet again.
The device should now boot straight to OpenWrt.

Signed-off-by: Tobias Schramm <tobleminer@gmail.com>

### CODE REVIEW

File: target/linux/ramips/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -262,6 +262,14 @@ m2m)
 	set_wifi_led "$boardname:blue:wifi"
 	ucidef_set_led_netdev "eth" "Ethernet" "$boardname:green:wan" "eth0"
 	;;
+mikrotik,rbm11g)
+	ucidef_set_rssimon "wlan0" "200000" "1"
+	ucidef_set_led_rssi "rssilow" "RSSILOW" "rbm11g:green:rssi0" "wlan0" "26" "100" "-25" "13"
```
Comment (mkresin) [inline target/linux/ramips/base-files/etc/board.d/01_leds]: Please use "$boardname:green:..." for the LEDs.

Comment (mkresin) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: Do you have pwm leds (LEDs which can have different brightness level)? If not the last two values can be omitted.

I'm somehow in doubt that you checked the rssi led syntax. It is:

`... <led> <wireless interface> <signal quality min> <signal quality max> ...`

With five rssi LEDs I rather would expect a configuration like:

```
ucidef_set_led_rssi "rssilow" "RSSILOW" "$boardname:green:rssi0" "wlan0" "1" "100"
ucidef_set_led_rssi "rssimediumlow" "RSSIMEDIUMLOW" "$boardname:green:rssi1" "wlan0" "20" "100"
ucidef_set_led_rssi "rssimedium" "RSSIMEDIUM" "$boardname:green:rssi2" "wlan0" "40" "100"
ucidef_set_led_rssi "rssimediumhigh" "RSSIMEDIUMHIGH" "$boardname:green:rssi3" "wlan0" "60" "100"
ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "$boardname:green:rssi4" "wlan0" "80" "100"
```

rssi0 is on if the quality is between 1% and 100%
rssi1 is on if the quality is between 20% and 100%

and so on.

File: target/linux/ramips/base-files/etc/board.d/02_network [hunk 2]
```diff
@@ -289,6 +289,10 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch1" \
 			"0:lan" "1:lan" "2:lan" "3:lan" "4:wan" "7t@eth0"
 		;;
+	mikrotik,rbm11g)
```
Comment (mkresin) [inline target/linux/ramips/base-files/etc/board.d/02_network]: A switch config for a single port device doesn't make that much sense. Please move the board to the block around line 40, which disables the switch and sets eth0 as lan interface.

File: target/linux/ramips/base-files/etc/diag.sh [hunk 3]
```diff
@@ -209,6 +209,9 @@ get_status_led() {
 	m4-8M)
 		status_led="m4:blue:status"
 		;;
+	mikrotik,rbm11g)
+		status_led="rbm11g:green:usr"
```
Comment (mkresin) [inline target/linux/ramips/base-files/etc/diag.sh]: Please use `$boardname:green:usr` and merge it with the rbm33g block

File: target/linux/ramips/dts/RBM11G.dts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+		};
+	};
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+		res {
+			label = "res";
```
Comment (mkresin) [inline target/linux/ramips/dts/RBM11G.dts]: do you mean reset by any chance?

File: target/linux/ramips/dts/RBM11G.dts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "uart2", "wdt", "sdhci";
+			ralink,function = "gpio";
+		};
+	};
+};
+
+&i2c {
+	status = "okay";
```
Comment (mkresin) [inline target/linux/ramips/dts/RBM11G.dts]: Out of curiosity, is something connected to the i2c bus?

File: target/linux/ramips/dts/RBM11G.dts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+&ethernet {
+	mtd-mac-address = <&factory 0x0010>;
+	mtd-mac-address-increment = <1>;
+};
+
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "uart2", "wdt", "sdhci";
```
Comment (mkresin) [inline target/linux/ramips/dts/RBM11G.dts]: The rssi leds are using the pins of the rgmii2 group, which is at least missing. Please use https://www.kresin.me/files/pinmux_review_ramips.ods to check which pin groups need to be set to the gpio mode/function.

**Review by mkresin — CHANGES_REQUESTED** [review]

File: target/linux/ramips/dts/RBM11G.dts (hunk 6 above)
Comment (TobleMiner) [reply target/linux/ramips/dts/RBM11G.dts]: Nice cheatsheet, thanks for sharing!

File: target/linux/ramips/base-files/etc/board.d/01_leds (hunk 1 above)
Comment (TobleMiner) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: Your doubts are well founded :smile: . I'll definetly fix that

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 2 above)
Comment (TobleMiner) [reply target/linux/ramips/base-files/etc/board.d/02_network]: Makes sense. Didn't notice the devices with disabled switch, thanks.

File: target/linux/ramips/base-files/etc/diag.sh (hunk 3 above)
Comment (TobleMiner) [reply target/linux/ramips/base-files/etc/diag.sh]: Yeah, didn't notice that $boardname != $board and thus thought that introducing special characters into the name would not be a good idea. But that thought didn't make a lot of sense. Thanks for clarifying.

File: target/linux/ramips/dts/RBM11G.dts (hunk 4 above)
Comment (TobleMiner) [reply target/linux/ramips/dts/RBM11G.dts]: Yep, simple typo. Thanks

File: target/linux/ramips/dts/RBM11G.dts (hunk 5 above)
Comment (TobleMiner) [reply target/linux/ramips/dts/RBM11G.dts]: My best guess would be that i2c is maybe connected to SMBUS on the miniPCIe slot. I'm not sure about that though. Thus I left i2c enabled.

**Comment by TobleMiner** [comment]
@mkresin I think I've implemented all of your suggestions. Thanks for reviewing!

**Pull request closed without merging**